src/Workspaces/Core/Portable/Remote/RemoteHostSessionHelpers.cs

```diff
@@ -155,104 +161,48 @@ public async Task<T> TryInvokeAsync<T>(string targetName, IReadOnlyList<object>
-        }
-
-        public async Task<bool> TryInvokeAsync(string targetName, Solution solution, IReadOnlyList<object> arguments, CancellationToken cancellationToken)
-        {
             using var pooledObject = SharedPools.Default<List<object>>().GetPooledObject();
-            using var scope = await solution.GetPinnedScopeAsync(cancellationToken).ConfigureAwait(false);
-            using var connection = await TryGetConnectionAsync(cancellationToken).ConfigureAwait(false);
-            if (connection == null)
-            {
-                return false;
-            }
-
             pooledObject.Object.Add(scope.SolutionInfo);
             pooledObject.Object.AddRange(arguments);
 
             await connection.Target.InvokeAsync(targetName, pooledObject.Object, cancellationToken).ConfigureAwait(false);
             return true;
         }
 
-        public async Task<T> TryInvokeAsync<T>(string targetName, Solution solution, IReadOnlyList<object> arguments, CancellationToken cancellationToken)
+        public async Task<Optional<T>> TryInvokeAsync<T>(string targetName, Solution solution, IReadOnlyList<object> arguments, CancellationToken cancellationToken)
         {
-            using var pooledObject = SharedPools.Default<List<object>>().GetPooledObject();
-            using var scope = await solution.GetPinnedScopeAsync(cancellationToken).ConfigureAwait(false);
             using var connection = await TryGetConnectionAsync(cancellationToken).ConfigureAwait(false);
             if (connection == null)
             {
                 return default;
             }
 
-            pooledObject.Object.Add(scope.SolutionInfo);
-            pooledObject.Object.AddRange(arguments);
-
-            return await connection.Target.InvokeAsync<T>(targetName, pooledObject.Object, cancellationToken).ConfigureAwait(false);
-        }
-
-        public async Task<bool> TryInvokeAsync(
-            string targetName, Solution solution, IReadOnlyList<object> arguments, Func<Stream, CancellationToken, Task> funcWithDirectStreamAsync, CancellationToken cancellationToken)
-        {
-            using var pooledObject = SharedPools.Default<List<object>>().GetPooledObject();
-            using var scope = await solution.GetPinnedScopeAsync(cancellationToken).ConfigureAwait(false);
-            using var connection = await TryGetConnectionAsync(cancellationToken).ConfigureAwait(false);
-            if (connection == null)
-            {
-                return false;
-            }
-
-            pooledObject.Object.Add(scope.SolutionInfo);
-            pooledObject.Object.AddRange(arguments);
-
-            await connection.Target.InvokeAsync(targetName, pooledObject.Object, funcWithDirectStreamAsync, cancellationToken).ConfigureAwait(false);
-            return true;
-        }
+            using var scope = await _remotableDataService.CreatePinnedRemotableDataScopeAsync(solution, cancellationToken).ConfigureAwait(false);
```

Comment:
for my own understanding. w hat's a 'pinned remotable data scope'?

Reply:
Keeps a group of objects alive in devenv.exe so they are accessible by an assigned ID for the duration of an asynchronous OOP call.